### ISSUE TITLE
Protocol not available issue on LX

### ISSUE THREAD

**Issue by blackwood821** [joyent/illumos-joyent#330]
I created a LX CentOS zone from the `centos-7@20180323` image and installed the `bind` package. The `bind` service fails to start and I see the following in `/var/named/data/named.run`:

```
managed-keys-zone: journal file is out of date: removing journal file
managed-keys-zone: loaded serial 14
zone 0.in-addr.arpa/IN: loaded serial 0
zone localhost/IN: loaded serial 0
zone 1.0.0.127.in-addr.arpa/IN: loaded serial 0
zone localhost.localdomain/IN: loaded serial 0
zone 1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.ip6.arpa/IN: loaded serial 0
all zones loaded
running
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(526, IPV6_RECVTCLASS) failed: Protocol not available
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(527, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:dc3::35#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(528, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:dc3::35#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(526, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:500:2d::d#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(527, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:500:2d::d#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(528, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:500:9f::42#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(526, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:500:9f::42#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(527, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:7fd::1#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(529, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:7fd::1#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(530, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:500:2f::f#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(531, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:500:2f::f#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(532, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:503:c27::2:30#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(533, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:503:c27::2:30#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(534, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:500:2::c#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(535, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:500:2::c#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(536, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:500:a8::e#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(537, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:500:a8::e#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(538, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:503:ba3e::2:30#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(539, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:503:ba3e::2:30#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(540, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:500:1::53#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(541, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:500:1::53#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(526, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:500:12::d0d#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(527, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:500:12::d0d#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(528, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:7fe::53#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(526, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:7fe::53#53
../../../../lib/isc/unix/socket.c:2891: unexpected error:
setsockopt(527, IPV6_RECVTCLASS) failed: Protocol not available
network unreachable resolving './DNSKEY/IN': 2001:500:200::b#53
../../../../lib/isc/unix/socket.c:2904: unexpected error:
setsockopt(528, IP_RECVTOS) failed: Protocol not available
network unreachable resolving './NS/IN': 2001:500:200::b#53
../../../../lib/isc/unix/socket.c:2904: unexpected error:
setsockopt(526, IP_RECVTOS) failed: Protocol not available
../../../../lib/isc/unix/socket.c:2904: unexpected error:
setsockopt(25, IP_RECVTOS) failed: Protocol not available
../../../../lib/isc/unix/socket.c:2904: unexpected error:
setsockopt(26, IP_RECVTOS) failed: Protocol not available
managed-keys-zone: Key 20326 for zone . acceptance timer complete: key now trusted
resolver priming query complete
```
I then tried disabling IPv6 since I only needed IPv4 by setting `OPTIONS="-4"` in `/etc/sysconfig/named`. Then I got a different error:

```
managed-keys-zone: journal file is out of date: removing journal file
managed-keys-zone: loaded serial 13
zone 0.in-addr.arpa/IN: loaded serial 0
zone localhost.localdomain/IN: loaded serial 0
zone localhost/IN: loaded serial 0
zone 1.0.0.127.in-addr.arpa/IN: loaded serial 0
zone 1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.ip6.arpa/IN: loaded serial 0
all zones loaded
running
../../../../lib/isc/unix/socket.c:2904: unexpected error:
setsockopt(519, IP_RECVTOS) failed: Protocol not available
../../../../lib/isc/unix/socket.c:2904: unexpected error:
setsockopt(520, IP_RECVTOS) failed: Protocol not available
../../../../lib/isc/unix/socket.c:2904: unexpected error:
setsockopt(23, IP_RECVTOS) failed: Protocol not available
../../../../lib/isc/unix/socket.c:2904: unexpected error:
setsockopt(24, IP_RECVTOS) failed: Protocol not available
managed-keys-zone: Key 20326 for zone . acceptance timer complete: key now trusted
resolver priming query complete
```

I switched over to native SmartOS zone and everything is working properly.

**Comment by danmcd** [comment]
Neither of these socket options are implemented even in native illumos today.

Fixing this for LX means also fixing it for illumos.  This issue may get linked with an upstream illumos bug.  Personally I can't drop what I'm doing and fix this, but before the week end's I'll triage this and file an accompanying bug for illumos.

**Comment by danmcd** [comment]
A check of the BIND source (specifically https://gitlab.isc.org/isc-projects/bind9/blob/a5a1cbece45e6ca68aafe3b9b995eac6b0f45dd2/lib/isc/unix/socket.c), indicates that BIND does or doesn't set these socket options if the option is defined in the header file.  For Linux, these most likely are, so we cannot constrain this fix to "returning the right errno" as some LX emulation does.

I've filed illumos#13175:  https://www.illumos.org/issues/13175  to track this in illumos.  That fix is required for this to be fixed in LX.  Ideally, whomever fixes 13175 should use SmartOS or OmniOSce (both have LX zones) for development, and make sure the LX changes are separable so they can land shortly after 13175 does.

**Comment by danmcd** [comment]
Turns out IPV6_RECVTCLASS is already in illumos, so we may merely need to expose it in LX.  The IP_RECVTOS, however, is still in the purview of illumos#13175.

**Comment by danmcd** [comment]
For IPV6_RECVTCLASS, it might be as simple as tweaking around here: https://grok.elemental.org/source/xref/illumos-joyent/usr/src/uts/common/brand/lx/syscall/lx_socket.c?r=b46b025a#2777

And funny thing, if you use an AF_INET6 socket for v4 and v6, you the IPV6_RECVTCLASS give you the v4 TOS for free (in a IPV6_RECVTCLAS ancillary message).  THAT might make it enough for BIND.  I can probably fix the IPV6 case quickly and build you a test PI if you're so inclined to test it out.

**Comment by danmcd** [comment]
And if you're really brave, here's a possible hotfix you can run in the global zone (but after which you'll have to restart bind):

`echo 'ltos_ipv6_sockopts+210/Z0x400000019' | mdb -kw`

**Comment by danmcd** [comment]
diff --git a/usr/src/uts/common/brand/lx/syscall/lx_socket.c b/usr/src/uts/common/brand/lx/syscall/lx_socket.c
index b67deee14f..642bdb9e6f 100644
--- a/usr/src/uts/common/brand/lx/syscall/lx_socket.c
+++ b/usr/src/uts/common/brand/lx/syscall/lx_socket.c
@@ -2774,7 +2774,7 @@ static const lx_sockopt_map_t ltos_ipv6_sockopts[LX_IPV6_TCLASS + 1] = {
        { OPTNOTSUP, 0 },
        { OPTNOTSUP, 0 },
        { OPTNOTSUP, 0 },
-       { OPTNOTSUP, 0 },                       /* IPV6_RECVTCLASS      */
+       { IPV6_RECVTCLASS, sizeof (int) },      /* IPV6_RECVTCLASS      */
        { IPV6_TCLASS, sizeof (int) }           /* IPV6_TCLASS          */
 };

**Comment by citrus-it** [comment]
> -       { OPTNOTSUP, 0 },                       /* IPV6_RECVTCLASS      */
> +       { IPV6_RECVTCLASS, sizeof (int) },      /* IPV6_RECVTCLASS      */

It's going to need adding to the cmsg translation table too.

**Comment by blackwood821** [comment]
> For IPV6_RECVTCLASS, it might be as simple as tweaking around here: https://grok.elemental.org/source/xref/illumos-joyent/usr/src/uts/common/brand/lx/syscall/lx_socket.c?r=b46b025a#2777
> 
> And funny thing, if you use an AF_INET6 socket for v4 and v6, you the IPV6_RECVTCLASS give you the v4 TOS for free (in a IPV6_RECVTCLAS ancillary message). THAT might make it enough for BIND. I can probably fix the IPV6 case quickly and build you a test PI if you're so inclined to test it out, @blackwood821

Thanks @danmcd. I switched to a native SmartOS zone when I encountered this issue so I'm in no rush but thanks for looking into this!

**Comment by jbarfield** [comment]
@danmcd could this bug also be related to the error below that I am getting below?

PI: joyent_20200813T030805Z

Trying to run elastiflow/logstash in a debian 9 LX zone:

[2020-11-23T21:00:35,248][ERROR][logstash.inputs.udp      ][elastiflow][769387adbe9162374989c7cf28dbfe8d169a5019d88caa0790c6ca870ea0c642] UDP listener died {:exception=>#<Errno::ENOPROTOOPT: Protocol not available - Protocol not available>, :backtrace=>["org/jruby/ext/socket/RubyBasicSocket.java:343:in `setsockopt'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-input-udp-3.3.4/lib/logstash/inputs/udp.rb:109:in `udp_listener'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-input-udp-3.3.4/lib/logstash/inputs/udp.rb:68:in `run'", "/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:378:in `inputworker'", "/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:369:in `block in start_input'"]}

**Comment by danmcd** [comment]
@jbarfield --> Seems that way.  Easiest way to find out might be to run a pi from release-20201022 or later.  The fix mentioned in #337 was integrated for that release.
